src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingWatcher.cs

```diff
@@ -84,43 +89,25 @@ private FileChangeList ComputeChangesAndUpdateState()
             _version++;
             var changes = new FileChangeList();
 
-            WIN32_FIND_DATAW fileData = new WIN32_FIND_DATAW();
-            unsafe
+            if (Directory.Exists(_directory))
             {
-                WIN32_FIND_DATAW* pFileData = &fileData;
-
-                for(int index=0; index < _directories.Count; index++) {
-                    var directory = _directories[index];
-
-                    var handle = DllImports.FindFirstFileExW(directory, FINDEX_INFO_LEVELS.FindExInfoBasic, pFileData, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, DllImports.FIND_FIRST_EX_LARGE_FETCH);
-                    if (handle == DllImports.INVALID_HANDLE_VALUE) { // directory got deleted 
-                        if (Tracing.Switch.ShouldTrace(TraceEventType.Information))
-                        {
-                            Tracing.TraceEvent(TraceEventType.Warning, 2, "Directory could not be opened {0}", directory);
-                        }
-                        _directories.Remove(directory);
-                        continue;
-                    }
-
-                    try
-                    {
-                        do
-                        {
-                            if (IsSpecial(fileData.cFileName)) continue;
-                            if (!IsWatched(fileData.cFileName)) continue;
-                            UpdateState(directory, ref changes, ref fileData, fileData.cFileName);
-                        }
-                        while (DllImports.FindNextFileW(handle, pFileData));
-                    }
-                    finally
-                    {
-                        DllImports.FindClose(handle);
-                    }
+                var files = new FileSystemEnumerable<FileInfo>(
+                    _directory,
+                    (ref FileSystemEntry entry) => (FileInfo)entry.ToFileSystemInfo(),
```

Comment:
This will allocate more than you need to. In the delegate you can call `UpdateState` with the `FileSystemEntry`. As that will capture `changes` the delegate won't get cached, however (resulting in an extra allocation). You can eliminate that by deriving directly from `FileSystemEnumerator<TResult>`. `FileSystemEnumerable` is the 80% case helper (it wraps `FileSystemEnumerator` and we use it internally for existing IO APIs). `FileSystemEnumerator` is the advanced, full control option. :)

Reply:
Thanks for helping me make my code awesome! Is this what you had in mind:

```csharp
var files = new FileSystemEnumerable<FileInfo>(
    _directory,
    (ref FileSystemEntry entry) => (FileInfo)entry.ToFileSystemInfo(),
    new EnumerationOptions { IgnoreInaccessible = true, RecurseSubdirectories = _includeSubdirectories })
{
    ShouldIncludePredicate = (ref FileSystemEntry entry) => IsWatched(ref entry)
};
var fileEnumerator = files.GetEnumerator();
while( fileEnumerator.MoveNext())
{
    UpdateState(_directory, ref changes, fileEnumerator.Current);
}
```

I don't feel like I gained anything. But you also said I should derive from `FileSystemEnumerator`, but I'm not really sure where to go with that. `FileSystemEnumaerable` seems to be tied to `FileSystemEnumerator`, so did you want me to derive from `FileSystemEnumerable`

I'm also not seeing the extra allocations that I need to get rid of, so all pointers are welcome. Thanks for your help!